### DISCUSSION TITLE
The pods are not in Running state

### DISCUSSION THREAD

**Discussion by jvaidyan** [sse-secure-systems/connaisseur#321] [Q&A] [answered]
I deployed the latest connaisseur helm chart in an openshift cluster. The Pods are not in running state.
The describe pod doesnt show much useful information other than the Readiness probe failed.
The logs also shows only the probe information.  How can I troubleshoot to see the underlying issue?

Logs
[2021-09-27 18:20:09,666] INFO: 10.131.0.1 - - [27/Sep/2021 18:20:09] "GET /ready HTTP/1.1" 500 -
[2021-09-27 18:20:09,877] INFO: 10.128.2.1 - - [27/Sep/2021 18:20:09] "GET /health HTTP/1.1" 200 -
[2021-09-27 18:20:10,721] INFO: 10.128.2.1 - - [27/Sep/2021 18:20:10] "GET /ready HTTP/1.1" 500 -
[2021-09-27 18:20:11,712] INFO: 10.129.2.1 - - [27/Sep/2021 18:20:11] "GET /ready HTTP/1.1" 500 -

$ kubectl get all -n connaisseur
NAME                                          READY   STATUS    RESTARTS   AGE
pod/connaisseur-deployment-5ff859845f-2rmwp   0/1     Running   0          105m
pod/connaisseur-deployment-5ff859845f-k8ndj   0/1     Running   0          105m
pod/connaisseur-deployment-5ff859845f-ths7h   0/1     Running   0          105m
pod/demo-595648d5f4-c7gtr                     1/1     Running   0          10m

NAME                      TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)   AGE
service/connaisseur-svc   ClusterIP   172.30.158.139   <none>        443/TCP   105m

NAME                                     READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/connaisseur-deployment   0/3     3            0           105m
deployment.apps/demo                     1/1     1            1           10m

NAME                                                DESIRED   CURRENT   READY   AGE
replicaset.apps/connaisseur-deployment-5ff859845f   3         3         0       105m
replicaset.apps/demo-595648d5f4                     1         1         1       10m

**Comment by jvaidyan** [comment, marked as answer]
The helm chart uses the default values. No overrides for the Notary server.

**Reply by jvaidyan** [reply]
To add, I did a helm upgrade to turn detectionMode: true as part of troubleshooting this issue.

**Reply by xopham** [reply]
Hi @jvaidyan! Thanks for reaching out.

The `helm upgrade` is currently non-functional due to a cert issue and we will provide a fix in the upcoming release within the next weeks, see e.g. #255 .
To fix the situation right now, I would recommend doing a `make annihilate` or manually run:
```bash
kubectl delete all,mutatingwebhookconfigurations,clusterroles,clusterrolebindings,configmaps,imagepolicies,secrets,serviceaccounts,crds -lapp.kubernetes.io/instance=connaisseur	kubectl delete ns connaisseur
```
Then install Connaisseur again.
Openshift might require some configuration adjustments by the way, as reported by #288

Let me know if that helps :pray:

**Reply by jvaidyan** [reply]
@xopham  Sure , I will try deleting and reinstalling. Because of policies I have removed the securityContext values in job, pod and deployment templates.  
The #288 mentions about a setting  openshift: true  . Is this functional in latest chart ?

**Reply by jvaidyan** [reply]
@xopham 
I add below in the deployment template as mentioned in #288 
and also the switch in values.yaml to have openshift: true
{{- if not (.Values.openshift) }}
      annotations:
        seccomp.security.alpha.kubernetes.io/pod: runtime/default
      {{- end }}

Deleted all resources and cleaned up the namespace. Ran the helm install again and the containers are not in Ready State (same as before).

**Reply by jvaidyan** [reply]
@xopham  When I reran the helm (after cleanup) , I containers get into ready state and stays until the bootstrap sentinel gets into completed status. I noticed below from events. The values.yaml doesnt contain a default value for these.

8m         Warning   FailedCreate        job/connaisseur-install-webhook-job           Error creating: pods "connaisseur-install-webhook-job-" is forbidden: unable to validate against any security context constraint: [spec.containers[0].securityContext.runAsUser: Invalid value: 10001: must be in the ranges: [1000740000, 1000749999]]

I repeated the helm install after removing the security context from job.yaml , but in this case the sentinel pod doesnt get  deployed

3m55s       Normal    Pulled                         pod/connaisseur-bootstrap-sentinel            Successfully pulled image "busybox:stable" in 2.062563974s
3m45s       Warning   Failed                         pod/connaisseur-bootstrap-sentinel            Error: container create failed: time="2021-09-27T23:37:29Z" level=error msg="container_linux.go:366: starting container process caused: process_linux.go:472: container init caused: read init-p: connection reset by peer"

I am not sure how these all are related. 
Bottomline, my issue is still not fixed.

**Reply by xopham** [reply]
@jvaidyan it seems to be an issue with the openshift compatibility. Could you remove the security context sections for the [deployment](https://github.com/sse-secure-systems/connaisseur/blob/master/helm/templates/deployment.yaml#L73), [bootstrap sentinel](https://github.com/sse-secure-systems/connaisseur/blob/master/helm/templates/bootstrap-sentinel.yaml#L21) and for [all 3 jobs](https://github.com/sse-secure-systems/connaisseur/blob/master/helm/templates/webhook/job.yaml)?

It seems that some part in the installation order does not startup properly.

**Reply by jvaidyan** [reply]
@xopham  , removed the securitycontext from deployment, sentinel pod and jobs. Still the same result.
Events show a failure related to the sentinel pod. What is the significance of this sentinel pod ?



137m        Warning   FailedCreatePodSandBox   pod/connaisseur-bootstrap-sentinel   Failed to create pod sandbox: rpc error: code = Unknown desc = `/usr/bin/runc --root /run/runc start 139b67aa7bdf4933b7fb85cbd69142acf8e8988a85f8aa506f18253b363b76d6` failed: time="2021-09-28T13:29:06Z" level=error msg="cannot start a container that has stopped"
  (exit status 1)
137m        Normal    Pulling                  pod/connaisseur-bootstrap-sentinel             Pulling image "busybox:stable"
137m        Normal    Pulled                   pod/connaisseur-bootstrap-sentinel             Successfully pulled image "busybox:stable" in 1.14800584s
136m        Warning   Failed                   pod/connaisseur-bootstrap-sentinel             Error: container create failed: time="2021-09-28T13:29:39Z" level=error msg="container_linux.go:366: starting container process caused: process_linux.go:472: container init caused: "
137m        Normal    Scheduled                pod/connaisseur-deployment-59f4d44cdc-2zzhb    Successfully assigned connaisseur/connaisseur-deployment-59f4d44cdc-2zzhb to contino-aro-cluster-x6kvf-worker-centralus2-8lt8t
:

**Reply by jvaidyan** [reply]
@xopham When I remove the readiness probe the containers are in Ready and running status. But I am not sure if this fixes all teh issues as I am able to perform kubectl run demo --image=docker.io/securesystemsengineering/testimage:unsigned

**Reply by xopham** [reply]
The sentinel is required to fix the boot order problem that is sort of a chicken-egg issue as described [in ADR 1](https://sse-secure-systems.github.io/connaisseur/v2.1.2/adr/ADR-1_bootstrap-sentinel/).
We found a way to solve this more gracefully and try to release the new version. It is described [in ADR 5](https://github.com/sse-secure-systems/connaisseur/pull/255/files#diff-240a0c8e71a253290bc78ba29922f0357ec4d7e50992a4119ab29137e19e0a31) which overall will improve the usability.

Still struggling with your error though. Can you get more specific information via `kubectl lods <pod name>`?

**Reply by xopham** [reply]
Maybe you cannot drop the full security context, but only change the group and user to `RunAsAny` or something. Unfortunately, I have limited experience with openshift. There is some docs on it though: https://docs.openshift.com/container-platform/4.1/authentication/managing-security-context-constraints.html

Would love to figure this out as we really value compatibility...

**Reply by jvaidyan** [reply]
@xopham  The logs for connaisseur pods only show the health and readiness probe messages,,Unable to check logs for sentinel pod as it doesnt come up .


         Warning   Failed                         pod/connaisseur-bootstrap-sentinel                  Error: container create failed: time="2021-09-29T13:55:03Z" level=error msg="container_linux.go:366: starting container process caused: process_linux.go:472: container init caused: read init-p: connection reset by peer"

**Reply by xopham** [reply]
Maybe it is unrelated to us. There is a very similar bug described for OCP 4.7: https://bugzilla.redhat.com/show_bug.cgi?id=1959322

Do you use a similar version?

**Reply by xopham** [reply]
I somehow can't seem to learn anything from the error message above. I suppose that is the `kubectl describe` of the sentinel, right? What does it show for the others?

**Reply by jvaidyan** [reply]
@xopham that was from the kubectl get events..The other events looks normal interms of deployment and jobs.

**Reply by xopham** [reply]
Hm. What does `kubectl describe` show?

Have once deleted all resources again and even removed the namespace?

It technically should work with Openshift / OKD 4.7, as reported by @pflaeging in #288

**Reply by jvaidyan** [reply]
@xopham Was not able to do a describe on sentinel pod as it never came up. Have cleaned up the resources , namespace and did fresh install , but same error.

**Reply by jvaidyan** [reply]
@xopham Is there an alternate way of install other than helm ?

**Reply by jvaidyan** [reply]
@xopham  Finally we got it to work, These were the steps and I have questions on these:
1) We had to add securitycontext policy to the service accounts . So we created the 2 service accounts up front and added the scc policy to those, before running helm install

2) This step, I am not sure what is the right approach. Following https://sse-secure-systems.github.io/connaisseur/v2.1.2/getting_started/ we generated the public key using **_docker trust key generate root_** and added that to values.yaml as default under validator. Is that needed?  Is this step needed for each cluster we deploy ?

Please validate these steps.

**Reply by xopham** [reply]
Awesome! Happy to hear that you got it to work :tada:
For your point 1, it sounds a bit like there is another Admission Controller enforcing pod security policies or something. Could that be?
To your point 2, the validator and image policy configuration is completely up to you and your usecase. The `default` validator is just a simple way to get started. To create your own specific configuration, I recommend to read [the basics](https://sse-secure-systems.github.io/connaisseur/v2.1.2/basics/) and [notary specific section](https://sse-secure-systems.github.io/connaisseur/v2.1.2/validators/notaryv1/).
For notary v1 validation, you can only setup one root key per image registry (account) and will have to use that for any Connaisseur instance verifying images within repositories of that account.

**Reply by jvaidyan** [reply]
@xopham The default configuration in values.yaml has a key as shown below..
Is that a valid key ?  What I am trying to understand is that , for getting the admission controller to work with docker hub images, do we need to generate the key specifically or will the default key provided in the values.yaml work?


name: dockerhub_basics
  type: notaryv1
  host: notary.docker.io
  trust_roots:
    # public key for official docker images (https://hub.docker.com/search?q=&type=image&image_filter=official)
    # !if not needed feel free to remove the key!
  - name: docker_official
    key: |
      -----BEGIN PUBLIC KEY-----
      MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEOXYta5TgdCwXTCnLU09W5T4M4r9f
      QQrqJuADP6U7g5r9ICgPSmZuRHP/1AYUfOQW3baveKsT969EfELKj1lfCA==
      -----END PUBLIC KEY-----

**Reply by xopham** [reply]
@jvaidyan the key provided in the default config for docker hub under `docker_official` is docker's official public key that is for example used to sign the docker official images (e.g. `docker.io/library/redis`). For your own images, you need to configure your own validator with a key generated for your docker repository as described in the docs:
- https://sse-secure-systems.github.io/connaisseur/v2.1.2/basics/#detailed-configuration
- https://sse-secure-systems.github.io/connaisseur/v2.1.2/validators/notaryv1/

**Reply by jvaidyan** [reply]
Thanks. 
Just in case there are users with similar issues in openshift, below are the changes we had to do to get this working

**Option1:**  Prior to helm install, create the namespace and service accounts upfront (for webhook one remember to add the annotations) and remove the service account templates from /helm/templates and /helm/templates/webhook

oc adm policy add-scc-to-user privileged -z connaisseur-serviceaccount -n connaisseur
oc adm policy add-scc-to-user privileged -z connaisseur-hook-serviceaccount -n connaisseur

**Option 2**: Include scc creation as part of helm install.
Add templates for scc to both the service accounts.

**_/templates/scc.yaml_**
apiVersion: security.openshift.io/v1
kind: SecurityContextConstraints
metadata:
  name: {{ .Chart.Name }}-serviceaccount-scc
  namespace: {{ .Release.Namespace }}
  labels:
    {{- include "helm.labels" . | nindent 4 }}
users:
- system:serviceaccount:{{ .Release.Namespace }}:{{ .Chart.Name }}-serviceaccount
- system:admin
- system:serviceaccount:openshift-infra:build-controller
....... **(rest attributes similar to scc of privileged)**


**_/templates/webhook/scc.yaml_**


apiVersion: security.openshift.io/v1
kind: SecurityContextConstraints
metadata:
  name: {{ .Chart.Name }}-hook-serviceaccount-scc
  namespace: {{ .Release.Namespace }}
  labels:
    {{- include "helm.labels" . | nindent 4 }}
  annotations:
    helm.sh/hook: post-install, pre-delete, pre-upgrade, post-upgrade
    helm.sh/hook-delete-policy: hook-succeeded, hook-failed
users:
- system:serviceaccount:{{ .Release.Namespace }}:{{ .Chart.Name }}-hook-serviceaccount
- system:admin
- system:serviceaccount:openshift-infra:build-controller
....... **(rest attributes similar to scc of privileged)**


Thanks

**Reply by xopham** [reply]
Thanks for sharing :pray: Appears we will need some work to improve the compatibility for Openshift. In your experience, how are k8s tools commonly made compatible for openshift?
Would you be willing to check our upcoming Connaisseur v2.2.0 for compatibility? We'll remove the sentinel and slim down the deployment.

**Reply by jvaidyan** [reply]
Sure . Will checkout the 2.2.0 version

**Reply by xopham** [reply]
Awesome. Will let you know

**Reply by xopham** [reply]
@jvaidyan 
We have implemented some major improvements in the upcoming release v2.2.0 that also aim to improve compatibility and resolve some of the installation issues. We still need to fix some minor problems that were found during testing already, but hope to release it within the next 2 weeks :partying_face: 

I prepared a release candidate for you that is not the final version, but would help us see whether it resolves your issues. To test it, please follow the steps below:

1. Switch to the [test/release-candidate-v2.2.0](https://github.com/sse-secure-systems/connaisseur/tree/test/release-candidate-v2.2.0) branch.
2. Make sure to remove all prior Connaisseur resources, e.g. `make annihilate` or follow [the docs on deleting](https://sse-secure-systems.github.io/connaisseur/develop/basics/#delete).
3. Depending on your Kubernetes version (e.g. <v1.19) or flavor (e.g. Openshift), adjust the [securityContext in `helm/values.yaml`](https://github.com/sse-secure-systems/connaisseur/blob/test/release-candidate-v2.2.0/helm/values.yaml#L30). The fields in questions contain comments on needed changes.
4. Install connaisseur, e.g. `make install` or follow [the docs on deploying](https://sse-secure-systems.github.io/connaisseur/develop/basics/#deploy). Connaisseur should come up much quicker than before.
5. Once Connaisseur has come up, `kubectl get all -n connaisseur` should show the resources as described [in the docs on checking](https://sse-secure-systems.github.io/connaisseur/develop/basics/#check).
6. Test Connaisseur using the following test images (notary v1 and cosign):
```bash
# cosign-based signatures
kubectl run cosigned --image=xoph/demo:signed   # is accepted
kubectl run counsigned --image=xoph/demo:unsigned   # is denied

# notary-based signatures
kubectl run signed --image=docker.io/securesystemsengineering/testimage:signed  # is accepted
kubectl run unsigned --image=docker.io/securesystemsengineering/testimage:unsigned  # is denied
```

Please let me know, if that works for you or whether you encounter any issues during installation. If something fails, please share your logs via `kubectl describe` and/or `kubectl logs` of failing components.

Hope this helps :slightly_smiling_face:

**Reply by mrjoelkamp** [reply]
@xopham the `test/release-candidate-v2.2.0` branch fixed installation errors for me when running on a local Kind cluster. Prior to this version, installation would fail.

**Reply by xopham** [reply]
@mrjoelkamp awesome. Thanks for letting us know :pray: 
We are currently merging the last few fixes and will probably release tomorrow :rocket:
